### PR TITLE
Add restart controller

### PR DESCRIPTION
**Pull request by sbernauer**
## Description
For https://github.com/stackabletech/issues/issues/200
Actual implementation copied from https://github.com/stackabletech/restart-controller
Integration-Test: https://github.com/stackabletech/integration-tests/pull/207

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: deploy/helm/commons-operator/templates/roles.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+metadata:
+  name: {{ .Release.Name }}-clusterrole
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - pods
+      - configmaps
+      - secrets
+      - secrets
```
Comment (maltesander) [inline deploy/helm/commons-operator/templates/roles.yaml:13]: duplicate

File: deploy/manifests/roles.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+metadata:
+  name: commons-operator-clusterrole
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - pods
+      - configmaps
+      - secrets
+      - secrets
```
Comment (maltesander) [inline deploy/manifests/roles.yaml:14]: duplicate

File: rust/operator-binary/src/restart_controller.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    ObjectHasNoNamespace { obj_ref: ObjectRef<DynamicObject> },
+    #[snafu(display("failed to apply update to {}", obj_ref))]
+    Apply {
+        source: kube::Error,
+        obj_ref: ObjectRef<DynamicObject>,
+    },
+    #[snafu(display("configmaps were not yet loaded"))]
+    ConfigMapsUninited,
+    #[snafu(display("secrets were not yet loaded"))]
+    SecretsUninited,
```
Comment (maltesander) [inline rust/operator-binary/src/restart_controller.rs:48]: Uninited is not really english. What about Uninitialized or NotLoaded?

File: rust/operator-binary/src/restart_controller.rs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+impl ReconcilerError for Error {
+    fn category(&self) -> &'static str {
+        ErrorDiscriminants::from(self).into()
+    }
+}
```
Comment (maltesander) [inline rust/operator-binary/src/restart_controller.rs:55]:
```suggestion
impl ReconcilerError for Error {
    fn category(&self) -> &'static str {
        ErrorDiscriminants::from(self).into()
    }

    fn secondary_object(&self) -> Option<ObjectRef<DynamicObject>> {
        match self {
            Error::ObjectHasNoNamespace { obj_ref, .. } => {
                Some(obj_ref.clone().erase())
            }
            Error::MissingDruidConnString => None,
            Error::Apply { obj_ref, .. } => Some(obj_ref.clone().erase()),
            Error::ConfigMapsUninited { .. } => None,
            Error::SecretsUninited { .. } => None,
        }
    }
}
```

File: rust/operator-binary/src/restart_controller.rs [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                            "metadata": {
+                                "annotations": annotations,
+                            },
+                        },
+                    },
+                }),
+            ),
+        )
+        .await
+        .unwrap();
```
Comment (maltesander) [inline rust/operator-binary/src/restart_controller.rs:302]: This may fail? Can we have proper error handling here?

**Review by maltesander — COMMENTED** [review]
Just some comments. Probably @teozkr needs to have another look :)

File: rust/operator-binary/src/restart_controller.rs (hunk 3 above)
Comment (sbernauer) [reply rust/operator-binary/src/restart_controller.rs]: `Uninitialized` is fine for me :+1:

File: rust/operator-binary/src/restart_controller.rs (hunk 4 above)
Comment (sbernauer) [reply rust/operator-binary/src/restart_controller.rs]: Thanks for the hint! I implemented a little bit different, hope that's ok

File: rust/operator-binary/src/restart_controller.rs (hunk 5 above)
Comment (sbernauer) [reply rust/operator-binary/src/restart_controller.rs:311]: I'm struggling to create the ApplySnafu error :/ Can you please give me a hint how to do that?
I have tried lot's of variants, this is my current approach
```
        .await
        .context(ApplySnafu {
            obj_ref: ObjectRef::from_obj(&sts)
        })?;
```

**Review by teozkr — APPROVED** [review]
LGTM for now, there's some documentation missing and generalization that we might want to do, but that's more of a problem that I need to get back to myself...

**Comment by sbernauer** [comment]
Thanks! Actually I intended to write some docs after this is merged. But for the generalization i think you're the best man!

**Comment by sbernauer** [comment]
There is no dedicated ticket for docs but the docs are tracked as part of https://github.com/stackabletech/issues/issues/200

**Pull request merged**